### PR TITLE
feat(workflows): add reusable-release-multi-ecosystem

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Summary

Adds `reusable-release-multi-ecosystem.yml`, a sibling of `reusable-release-version-pr` for repos that bump multiple ecosystem manifests in one version PR (reference consumer: turbo-themes). Reuses existing `scripts/ci/release/*` plumbing with a new file→kind manifests runner.

### Decision gate (extend vs new)

**New sibling workflow**, not an overload of `reusable-release-version-pr.yml`.

- Existing version-pr is ecosystem-CSV + layout defaults (`node,ruby,python`, …).
- Multi-ecosystem needs a required `manifests` JSON map (`npm|raw|gemspec|pep621`), `bump` (`auto-from-commits` | `explicit`), and `prerelease-tag`.
- Sharing `calculate-version`, changelog, App-token PR creation, merge-queue skip, and failure reporting keeps one release family; a separate workflow keeps the consumer contracts clear.

## Type of Change

- [x] New feature (composite action, reusable workflow, or utility script)
- [ ] Bug fix
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD improvement

## Checklist

- [x] I have tested these changes locally
- [x] I have updated documentation if needed
- [x] Shell scripts pass `shellcheck`
- [x] YAML files pass `yamllint`
- [x] Python code passes `ruff` and `black`

## Breaking Changes

None

## Closes

- Closes #523
- Relates to #520

## Changes Made

- New `reusable-release-multi-ecosystem.yml` (`workflow_call`) with `manifests`, `bump`, `version`, `prerelease-tag`, `changelog`, `job-name`, egress inputs, and App-token secrets
- Kind updaters under `scripts/ci/release/ecosystems/`: `npm`, `raw`, `gemspec`, `pep621`, plus `_manifests_runner.sh`
- `resolve-multi-ecosystem-version.sh` wrapping `calculate-version.sh` with explicit bump + prerelease support
- Contract docs in `docs/workflow-contract.md` § Multi-ecosystem release contract; examples in `docs/reusable-workflows.md` and publishing index
- BATS coverage for each kind (rewrite, idempotency, prerelease, unknown-kind rejection) and workflow contract checks

## Testing

- [x] Full `uv run lintro chk` (0 issues)
- [x] Full BATS suite (`bats --recursive tests/bats/`, 2989 ok)
- [x] Script coverage ratchet satisfied
- [ ] No consumer migration in this PR (turbo-themes follows later per epic #520)

## Quality Checks

- [x] Lint / format clean
- [x] Signed commits
- [x] Action pins use full SHAs

## Notes

- Kind `pep621` may need PyPI endpoints in `allowed-endpoints` under `egress-policy: block` (same as existing `python.sh` tomlkit install).
- Egress harden ordering matches `reusable-release-version-pr` (two-phase checkout).

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds `reusable-release-multi-ecosystem.yml`, a sibling of `reusable-release-version-pr` for repos that need to bump multiple ecosystem manifest files (npm, raw, gemspec, pep621) in a single version PR. It introduces the manifests runner and four kind scripts, reusing the existing guard/changelog/App-token/merge-queue/failure-reporting plumbing.

- **New workflow** (`reusable-release-multi-ecosystem.yml`) correctly mirrors the sibling workflow's two-phase checkout, concurrency group, step-security harden runner, App-token wiring, and failure-reporting follow-up job.
- **Kind scripts** (`npm.sh`, `raw.sh`, `gemspec.sh`, `pep621.sh`) each validate their target file exists, perform an atomic write, and verify the result; the runner (`_manifests_runner.sh`) collects all failures before exiting non-zero so all errors are reported in one run.
- **`resolve-multi-ecosystem-version.sh`** wraps `calculate-version.sh` in an isolated subshell and adds explicit-bump and prerelease-tag support on top; BATS coverage includes the subshell isolation, v-prefix stripping, double-prerelease guard, and auto-from-commits with no releasable history.
</details>

<h3>Confidence Score: 5/5</h3>

Safe to merge; the workflow and scripts are well-structured, mirror the established sibling workflow exactly, and have comprehensive BATS coverage.

The implementation faithfully replicates the release-version-pr family's guard/changelog/App-token/merge-queue/failure-reporting pattern. Shell scripts use set -euo pipefail, quote all variables, validate inputs early, and verify writes atomically. The two advisory findings (SemVer vs PEP 440 prerelease format for the pep621 kind, and pep621 BATS tests skipping on clean runners) do not affect correctness for callers who supply PEP 440-compliant versions directly.

pep621.sh and the pep621-related BATS tests warrant a second look if Python package publishing via pyproject.toml is an intended use case for prerelease builds.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| .github/workflows/reusable-release-multi-ecosystem.yml | New reusable workflow with 603 lines; correctly mirrors reusable-release-version-pr structure with two-phase checkout, App-token PR creation, merge-queue skip guard, and failure-reporting follow-up job. |
| scripts/ci/release/ecosystems/_manifests_runner.sh | Orchestrates file-to-kind dispatching; validates JSON structure, entry count, kind allowlist, and script existence before running. Continues on individual failures to report all errors before exiting non-zero. |
| scripts/ci/release/ecosystems/pep621.sh | Updates [project].version via existing Python helpers. SemVer prerelease format (1.2.3-rc.1) produced by prerelease-tag input is not PEP 440 compliant, which could cause issues for Python package publishing. |
| scripts/ci/release/resolve-multi-ecosystem-version.sh | Wraps calculate-version.sh in an isolated subshell to capture outputs, then applies PRERELEASE_TAG with double-application guard and full semver validation. |
| tests/bats/unit/release/test_manifest_kind_updaters.bats | Good coverage for raw, npm, gemspec, and runner edge cases. All pep621 tests skip when tomlkit is absent, meaning they may never run in a clean CI environment despite pep621.sh installing it on demand. |

</details>

<details><summary><h3>Sequence Diagram</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
sequenceDiagram
    participant Caller as Caller Workflow
    participant HR as Harden Runner
    participant Tooling as lgtm-ci Tooling
    participant AppToken as create-github-app-token
    participant Guard as guard-release-commit.sh
    participant CheckPR as check-existing-version-pr.sh
    participant ResolveVer as resolve-multi-ecosystem-version.sh
    participant CalcVer as calculate-version.sh
    participant ManifestRunner as _manifests_runner.sh
    participant KindScript as npm/raw/gemspec/pep621
    participant CreatePR as create-pull-request

    Caller->>HR: egress-policy + allowed-endpoints
    Caller->>Tooling: sparse-checkout egress tooling only
    Caller->>Tooling: resolve-egress-allowlist composite
    Caller->>AppToken: RELEASE_APP_ID + RELEASE_APP_PRIVATE_KEY
    AppToken-->>Caller: app-token
    Caller->>Tooling: sparse-checkout scripts/ci/ with app-token
    Caller->>Guard: guard-release-commit.sh
    alt is release commit
        Guard-->>Caller: "is-release-commit=true, skip rest"
    else normal commit
        Caller->>CheckPR: check-existing-version-pr.sh
        alt PR already open
            CheckPR-->>Caller: "pr-exists=true, skip rest"
        else no existing PR
            Caller->>ResolveVer: BUMP_MODE + EXPLICIT_VERSION + PRERELEASE_TAG
            alt "bump=auto-from-commits"
                ResolveVer->>CalcVer: "subshell, GITHUB_OUTPUT=tmpfile"
                CalcVer-->>ResolveVer: next-version, bump-type, release-needed
            else "bump=explicit"
                Note over ResolveVer: Validate EXPLICIT_VERSION, strip v prefix
            end
            Note over ResolveVer: Append PRERELEASE_TAG if release-needed
            ResolveVer-->>Caller: next-version, bump-type, release-needed
            alt "release-needed=true"
                Caller->>ManifestRunner: NEXT_VERSION + MANIFESTS JSON map
                loop each path in manifests
                    ManifestRunner->>KindScript: MANIFEST_PATH + NEXT_VERSION
                    KindScript-->>ManifestRunner: "success or FAILED=1"
                end
                ManifestRunner-->>Caller: exit 0 all ok or exit 1 on any failure
                Caller->>CreatePR: branch, title, body with app-token
                CreatePR-->>Caller: PR URL + number
                Caller->>Caller: request CODEOWNER review
                Caller->>Caller: enable auto-merge if configured
            end
        end
    end
    Caller->>Caller: write summary always
    alt version-pr job failed
        Caller->>Caller: report-release-failure job opens GitHub issue
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
sequenceDiagram
    participant Caller as Caller Workflow
    participant HR as Harden Runner
    participant Tooling as lgtm-ci Tooling
    participant AppToken as create-github-app-token
    participant Guard as guard-release-commit.sh
    participant CheckPR as check-existing-version-pr.sh
    participant ResolveVer as resolve-multi-ecosystem-version.sh
    participant CalcVer as calculate-version.sh
    participant ManifestRunner as _manifests_runner.sh
    participant KindScript as npm/raw/gemspec/pep621
    participant CreatePR as create-pull-request

    Caller->>HR: egress-policy + allowed-endpoints
    Caller->>Tooling: sparse-checkout egress tooling only
    Caller->>Tooling: resolve-egress-allowlist composite
    Caller->>AppToken: RELEASE_APP_ID + RELEASE_APP_PRIVATE_KEY
    AppToken-->>Caller: app-token
    Caller->>Tooling: sparse-checkout scripts/ci/ with app-token
    Caller->>Guard: guard-release-commit.sh
    alt is release commit
        Guard-->>Caller: "is-release-commit=true, skip rest"
    else normal commit
        Caller->>CheckPR: check-existing-version-pr.sh
        alt PR already open
            CheckPR-->>Caller: "pr-exists=true, skip rest"
        else no existing PR
            Caller->>ResolveVer: BUMP_MODE + EXPLICIT_VERSION + PRERELEASE_TAG
            alt "bump=auto-from-commits"
                ResolveVer->>CalcVer: "subshell, GITHUB_OUTPUT=tmpfile"
                CalcVer-->>ResolveVer: next-version, bump-type, release-needed
            else "bump=explicit"
                Note over ResolveVer: Validate EXPLICIT_VERSION, strip v prefix
            end
            Note over ResolveVer: Append PRERELEASE_TAG if release-needed
            ResolveVer-->>Caller: next-version, bump-type, release-needed
            alt "release-needed=true"
                Caller->>ManifestRunner: NEXT_VERSION + MANIFESTS JSON map
                loop each path in manifests
                    ManifestRunner->>KindScript: MANIFEST_PATH + NEXT_VERSION
                    KindScript-->>ManifestRunner: "success or FAILED=1"
                end
                ManifestRunner-->>Caller: exit 0 all ok or exit 1 on any failure
                Caller->>CreatePR: branch, title, body with app-token
                CreatePR-->>Caller: PR URL + number
                Caller->>Caller: request CODEOWNER review
                Caller->>Caller: enable auto-merge if configured
            end
        end
    end
    Caller->>Caller: write summary always
    alt version-pr job failed
        Caller->>Caller: report-release-failure job opens GitHub issue
    end
```

</a>
</details>

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `scripts/ci/release/ecosystems/pep621.sh`, line 1001-1004 ([link](https://github.com/lgtm-hq/lgtm-ci/blob/fce04260743707bc7aa0430c15ec177662ebc7ce/scripts/ci/release/ecosystems/pep621.sh#L1001-L1004)) 

   <a href="#">P2</a> **Broad tomlkit version range without hash pinning**

   `python3 -m pip install --quiet 'tomlkit>=0.13,<1'` installs any tomlkit release in a half-open range spanning several minor versions, with no `--require-hashes` or lockfile. A future release within that range could silently change TOML serialization behaviour (e.g. key ordering, trailing newlines) and produce commits whose TOML is technically valid but differs from what callers expect. Consider pinning to an exact version or using `--require-hashes` with a known hash.

   Note: If this suggestion doesn't match your team's coding style, reply to this and let me know. I'll remember it for next time!

   <a href="https://app.greptile.com/api/ide/cursor?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20scripts%2Fci%2Frelease%2Fecosystems%2Fpep621.sh%0ALine%3A%201001-1004%0A%0AComment%3A%0A**Broad%20tomlkit%20version%20range%20without%20hash%20pinning**%0A%0A%60python3%20-m%20pip%20install%20--quiet%20'tomlkit%3E%3D0.13%2C%3C1'%60%20installs%20any%20tomlkit%20release%20in%20a%20half-open%20range%20spanning%20several%20minor%20versions%2C%20with%20no%20%60--require-hashes%60%20or%20lockfile.%20A%20future%20release%20within%20that%20range%20could%20silently%20change%20TOML%20serialization%20behaviour%20%28e.g.%20key%20ordering%2C%20trailing%20newlines%29%20and%20produce%20commits%20whose%20TOML%20is%20technically%20valid%20but%20differs%20from%20what%20callers%20expect.%20Consider%20pinning%20to%20an%20exact%20version%20or%20using%20%60--require-hashes%60%20with%20a%20known%20hash.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&pr=597&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursorDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCursor.svg?v=6">Fix in Cursor</picture></a> <a href="https://app.greptile.com/ide/claude-code?prompt=This%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20scripts%2Fci%2Frelease%2Fecosystems%2Fpep621.sh%0ALine%3A%201001-1004%0A%0AComment%3A%0A**Broad%20tomlkit%20version%20range%20without%20hash%20pinning**%0A%0A%60python3%20-m%20pip%20install%20--quiet%20'tomlkit%3E%3D0.13%2C%3C1'%60%20installs%20any%20tomlkit%20release%20in%20a%20half-open%20range%20spanning%20several%20minor%20versions%2C%20with%20no%20%60--require-hashes%60%20or%20lockfile.%20A%20future%20release%20within%20that%20range%20could%20silently%20change%20TOML%20serialization%20behaviour%20%28e.g.%20key%20ordering%2C%20trailing%20newlines%29%20and%20produce%20commits%20whose%20TOML%20is%20technically%20valid%20but%20differs%20from%20what%20callers%20expect.%20Consider%20pinning%20to%20an%20exact%20version%20or%20using%20%60--require-hashes%60%20with%20a%20known%20hash.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=lgtm-hq%2Flgtm-ci&pr=597&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaudeDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInClaude.svg?v=6">Fix in Claude Code</picture></a> <a href="https://app.greptile.com/api/ide/codex?prompt=IMPORTANT%3A%20Work%20in%20the%20repository%20%22lgtm-hq%2Flgtm-ci%22%20on%20the%20existing%20branch%20%22feat%2F523-reusable-release-multi-ecosystem%22.%20Checkout%20that%20branch%20%E2%80%94%20do%20NOT%20create%20a%20new%20branch%20or%20open%20a%20new%20PR.%20Push%20your%20changes%20to%20%22feat%2F523-reusable-release-multi-ecosystem%22.%0A%0AThis%20is%20a%20comment%20left%20during%20a%20code%20review.%0APath%3A%20scripts%2Fci%2Frelease%2Fecosystems%2Fpep621.sh%0ALine%3A%201001-1004%0A%0AComment%3A%0A**Broad%20tomlkit%20version%20range%20without%20hash%20pinning**%0A%0A%60python3%20-m%20pip%20install%20--quiet%20'tomlkit%3E%3D0.13%2C%3C1'%60%20installs%20any%20tomlkit%20release%20in%20a%20half-open%20range%20spanning%20several%20minor%20versions%2C%20with%20no%20%60--require-hashes%60%20or%20lockfile.%20A%20future%20release%20within%20that%20range%20could%20silently%20change%20TOML%20serialization%20behaviour%20%28e.g.%20key%20ordering%2C%20trailing%20newlines%29%20and%20produce%20commits%20whose%20TOML%20is%20technically%20valid%20but%20differs%20from%20what%20callers%20expect.%20Consider%20pinning%20to%20an%20exact%20version%20or%20using%20%60--require-hashes%60%20with%20a%20known%20hash.%0A%0AHow%20can%20I%20resolve%20this%3F%20If%20you%20propose%20a%20fix%2C%20please%20make%20it%20concise.&repo=lgtm-hq%2Flgtm-ci&pr=597&platform=github"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodexDark.svg?v=6"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/FixInCodex.svg?v=6">Fix in Codex</picture></a>
</details>

<sub>Reviews (2): Last reviewed commit: ["fix(release): address Greptile P2 review..."](https://github.com/lgtm-hq/lgtm-ci/commit/5178adb8b118c1f572f2fffa71d08cb49fc2aa05) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=44968379)</sub>

### CODE REVIEW

File: scripts/ci/release/ecosystems/npm.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+if [[ ! -f "$MANIFEST_PATH" ]]; then
+	log_error "[npm] package.json not found at: $MANIFEST_PATH"
+	exit 1
+fi
+
+export ECOSYSTEM_CONFIG_JSON
+ECOSYSTEM_CONFIG_JSON=$(jq -nc --arg p "$MANIFEST_PATH" '{package: $p}')
+
+log_info "[npm] Delegating to node updater for $MANIFEST_PATH"
+"$SCRIPT_DIR/node.sh"
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply scripts/ci/release/ecosystems/npm.sh:37]: Fixed: `npm.sh` now checks that `node.sh` exists and emits a clear `[npm]`-prefixed error before delegating.

File: .github/workflows/reusable-release-multi-ecosystem.yml [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+
+            ### Changelog
+
+            ${{ steps.changelog.outputs.changelog || '_Changelog generation disabled._' }}
+
+            ### What happens when this PR is merged?
+
+            1. The updated version files and changelog land on the target branch
+            2. If a separate auto-tag workflow is configured, it can create
+               `${{ inputs.tag-prefix }}${{ steps.version.outputs.next-version }}`,
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply .github/workflows/reusable-release-multi-ecosystem.yml:448]: Fixed: fallback text is now `_No changelog entries._`, which covers both `changelog: false` and an empty generated changelog.
